tests/msc2716_test.go

```diff
@@ -0,0 +1,538 @@
+				return r.Get("event_id").Str == eventWeShouldSee
+			})
+		})
+
+		t.Run("Unrecognised prev_event ID will throw an error", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+
+			backfillBulkHistoricalMessagesInReverseChronologicalAtTime(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				"$some-non-existant-event-id",
+				time.Now(),
+				1,
+				// Status
+				// TODO: Seems like this makes more sense as a 404
+				// But the current Synapse code around unknown prev events will throw ->
+				// `403: No create event in auth events`
+				403,
+			)
+		})
+
+		t.Run("Normal users aren't allowed to backfill messages", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			backfillBulkHistoricalMessagesInReverseChronologicalAtTime(
+				t,
+				alice,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				1,
+				// Status
+				// Normal user alice should not be able to backfill messages
+				403,
+			)
+		})
+
+		t.Run("Historical messages are visible on federated server", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			// Register and join the virtual user
+			ensureRegistered(t, as, virtualUserLocalpart)
+
+			backfillRes := backfillBulkHistoricalMessagesInReverseChronologicalAtTime(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				1,
+				// Status
+				200,
+			)
+			_, historicalEvents := getEventsFromBulkSendResponse(t, backfillRes)
+
+			// Join the room from a remote homeserver
+			remoteCharlie.JoinRoom(t, roomID, []string{"hs1"})
+
+			messagesRes := remoteCharlie.MustDoRaw(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, nil, "application/json", url.Values{
+				"dir":   []string{"b"},
+				"limit": []string{"100"},
+			})
+			messsageResBody := client.ParseJSON(t, messagesRes)
+			eventIDsFromResponse := getEventIDsFromResponseBody(t, messsageResBody)
+			// Since the original body can only be read once, create a new one from the body bytes we just read
+			messagesRes.Body = ioutil.NopCloser(bytes.NewBuffer(messsageResBody))
+
+			logrus.WithFields(logrus.Fields{
+				"eventIDsFromResponse": eventIDsFromResponse,
+				"historicalEvents":     historicalEvents,
+			}).Error("can we see historical?")
+
+			must.MatchResponse(t, messagesRes, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONCheckOffAllowUnwanted("chunk", []interface{}{historicalEvents[0]}, func(r gjson.Result) interface{} {
+						return r.Get("event_id").Str
+					}, nil),
+				},
+			})
```

Comment:
The historical message isn't being picked up  when `hs2` asks `hs1 to `/backfill` 🤔 

`hs2`:
```
2021-04-27 01:54:05,581 - synapse.access.http.8008 - 372 - INFO - POST-1 - ::ffff:172.25.0.1 - 8008 - {@charlie:hs2} Processed request: 0.815sec/-0.000sec (0.013sec, 0.006sec) (0.010sec/0.020sec/19) 37B 200 "POST /_matrix/client/r0/join/%21seVQrARDwefhpTBjhK:hs1?access_token=<redacted>&server_name=hs1 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-04-27 01:54:05,583 - synapse.metrics - 587 - INFO - sentinel - Collecting gc 1
2021-04-27 01:54:05,590 - synapse.appservice.scheduler - 85 - INFO - as_scheduler-0 - Starting appservice scheduler
2021-04-27 01:54:05,633 - synapse.federation.sender.transaction_manager - 119 - INFO - Measure[_send_new_transaction]@7fa8bb59ed60 - TX [hs1] {1619488442382} Sending transaction [1619488442382], (PDUs: 0, EDUs: 1)
2021-04-27 01:54:05,637 - synapse.http.matrixfederationclient - 209 - INFO - GET-4 - {GET-O-4} [hs1] Completed request: 200 OK in 0.01 secs - GET matrix://hs1/_matrix/federation/v1/backfill/%21seVQrARDwefhpTBjhK%3Ahs1?v=%241k2WeJtWJGZP53kghBe0u_iEacWQZ_xBM3Waa3wKrI4&limit=100
2021-04-27 01:54:05,657 - synapse.handlers.federation - 959 - INFO - GET-4 - backfill: Got 7 events with 0 edges
2021-04-27 01:54:05,699 - synapse.http.matrixfederationclient - 209 - INFO - Measure[_send_new_transaction]@7fa8bb59ed60 - {PUT-O-5} [hs1] Completed request: 200 OK in 0.07 secs - PUT matrix://hs1/_matrix/federation/v1/send/1619488442382
2021-04-27 01:54:05,700 - synapse.federation.sender.transaction_manager - 159 - INFO - Measure[_send_new_transaction]@7fa8bb59ed60 - TX [hs1] {1619488442382} got 200 response
2021-04-27 01:54:05,778 - synapse.access.http.8008 - 372 - INFO - GET-4 - ::ffff:172.25.0.1 - 8008 - {@charlie:hs2} Processed request: 0.187sec/-0.000sec (0.014sec, 0.011sec) (0.082sec/0.033sec/39) 2899B 200 "GET /_matrix/client/r0/rooms/%21seVQrARDwefhpTBjhK:hs1/messages?access_token=<redacted>&dir=b&limit=100 HTTP/1.1" "Go-http-client/1.1" [5 dbevts]
```

Reply:
See https://github.com/matrix-org/synapse/pull/9247#discussion_r620812210

This federation test is passing now! 🎉